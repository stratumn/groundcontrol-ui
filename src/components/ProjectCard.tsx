// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import graphql from "babel-plugin-relay/macro";
import React from "react";
import { createFragmentContainer } from "react-relay";
import {
  Button,
  Card,
  Dimmer,
  Divider,
  Header,
  Label,
  Loader,
 } from "semantic-ui-react";

import { ProjectCard_item } from "./__generated__/ProjectCard_item.graphql";

import CommitFeed from "./CommitFeed";
import RepositoryShortName from "./RepositoryShortName";

import "./ProjectCard.css";

export interface IProps {
  item: ProjectCard_item;
  onClone: (values: IProps) => any;
  onPull: (values: IProps) => any;
}

export function ProjectCard(props: IProps) {
  const {
    item,
    item: {
      repository,
      remoteReferenceShort,
      description,
      remoteCommits,
      isCloned,
      isCloning,
      isPulling,
      isAhead,
      isBehind,
      localReferenceShort,
    },
    onClone,
    onPull,
  } = props;
  const commitNodes = item.remoteCommits.edges.map(({ node }) => node);
  const labels: JSX.Element[] = [];
  const buttons: JSX.Element[] = [];
  const handleClone = () => onClone({ ...props });
  const handlePull = () => onPull({ ...props });
  const reference = remoteReferenceShort === localReferenceShort ?
    remoteReferenceShort : `${localReferenceShort} Â» ${remoteReferenceShort}`;

  let color: "grey" | "teal" | "pink" | "purple" = "grey";

  if (isCloned) {
    labels.push((
      <Label
        key="cloned"
        content="cloned"
        color="blue"
        size="small"
      />
    ));

    buttons.push((
      <Button
        key="pull"
        content="Pull"
        color="teal"
        icon="download"
        disabled={isPulling || !isBehind}
        loading={isPulling}
        onClick={handlePull}
      />
    ));

    if (!isBehind && !isAhead) {
      color = "teal";

      labels.push((
        <Label
          key="uptodate"
          content="up-to-date"
          color="teal"
          size="small"
        />
      ));
    }
  } else {
    buttons.push((
      <Button
        key="clone"
        content="Clone"
        color="teal"
        icon="clone"
        disabled={isCloning}
        loading={isCloning}
        onClick={handleClone}
      />
    ));
  }

  if (isAhead) {
    color = "purple";

    labels.push((
      <Label
        key="ahead"
        content="ahead"
        color="purple"
        size="small"
      />
    ));
  }

  if (isBehind) {
    color = "pink";

    labels.push((
      <Label
        key="behind"
        content="behind"
        color="pink"
        size="small"
      />
    ));
  }

  return (
    <Card
      className="ProjectCard"
      color={color}
    >
      <Card.Content>
        <Card.Header>
          <RepositoryShortName repository={repository} />
        </Card.Header>
        <Label size="small">{reference}</Label>
        {labels}
        <Card.Description>
          {description || "No description."}
        </Card.Description>
        <Divider horizontal={true}>
          <Header as="h6">Latest Commits</Header>
        </Divider>
        <CommitFeed items={commitNodes} />
      </Card.Content>
      <Card.Content extra={true}>
        <div className="ui three buttons">
          {buttons}
        </div>
      </Card.Content>
      <Dimmer
        active={remoteCommits.edges.length < 1}
        inverted={true}
      >
        <Loader content="Loading project commits..." />
      </Dimmer>
    </Card>
  );
}

export default createFragmentContainer(ProjectCard, graphql`
  fragment ProjectCard_item on Project
    @argumentDefinitions(
      commitsLimit: { type: "Int", defaultValue: 3 },
    ) {
    id
    repository
    remoteReferenceShort
    localReferenceShort
    description
    isCloning
    isCloned
    isPulling
    isBehind
    isAhead
    remoteCommits(first: $commitsLimit) {
      edges {
        node {
          ...CommitFeed_items
        }
      }
    }
  }`,
);
