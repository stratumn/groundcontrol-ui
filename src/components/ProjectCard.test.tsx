// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";
import { mocked } from "ts-jest/utils";

import { ProjectCard_item } from "./__generated__/ProjectCard_item.graphql";

import { mockQueryPropAttrs } from "../testing/relay";

import { IProps, ProjectCard } from "./ProjectCard";

jest.mock("./RepositoryShortName", () => "RepositoryShortName");

const item: ProjectCard_item = {
  ...mockQueryPropAttrs(),
  description: "description",
  id: "id",
  isAhead: false,
  isBehind: false,
  isCloned: false,
  isCloning: false,
  isPulling: false,
  localReferenceShort: "localReferenceShort",
  remoteCommits: {
    edges: [{
      node: {
        ...mockQueryPropAttrs(),
      },
    }, {
      node: {
        ...mockQueryPropAttrs(),
      },
    }],
  },
  remoteReferenceShort: "remoteReferenceShort",
  repository: "repository",
};

const props = {
  item,
  onClone: jest.fn(),
  onPull: jest.fn(),
};

beforeEach(() => {
  mocked(props.onClone).mockClear();
  mocked(props.onPull).mockClear();
});

describe("<ProjectCard />", () => {

  it("renders correctly", () => {
    const wrapper = shallow(<ProjectCard {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("shows a clone button if the project isn't cloned", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: false, isCloning: false } }}
      />,
    );
    expect(wrapper.find("Button[icon='clone'][loading=false]")).toHaveLength(1);
  });

  it("shows a loading clone button if the project is cloning", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: false, isCloning: true } }}
      />,
    );
    expect(wrapper.find("Button[icon='clone'][loading=true]")).toHaveLength(1);
  });

  it("shows a pull button if the project is cloned and behind", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isBehind: true } }}
      />,
    );
    expect(wrapper.find("Button[icon='download'][disabled=false]")).toHaveLength(1);
  });

  it("shows a disabled pull button if the project is cloned and up-to-date", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isBehind: false } }}
      />,
    );
    expect(wrapper.find("Button[icon='download'][disabled=true]")).toHaveLength(1);
  });

  it("shows a loading pull button if the project is pulling", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isPulling: true } }}
      />,
    );
    expect(wrapper.find("Button[icon='download'][loading=true]")).toHaveLength(1);
  });

  it("shows a cloned label if the project is cloned", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true } }}
      />,
    );
    expect(wrapper.find("Label[content='cloned']")).toHaveLength(1);
  });

  it("shows an up-to-date label if the project is up-to-date", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isBehind: false, isAhead: false } }}
      />,
    );
    expect(wrapper.find("Label[content='up-to-date']")).toHaveLength(1);
  });

  it("shows a behind label if the project is behind", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isBehind: true } }}
      />,
    );
    expect(wrapper.find("Label[content='behind']")).toHaveLength(1);
  });

  it("shows an ahead label if the project is ahead", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isAhead: true } }}
      />,
    );
    expect(wrapper.find("Label[content='ahead']")).toHaveLength(1);
  });

  it("triggers onClone when the clone button is clicked", () => {
    const cloneProps: IProps = {...props, item: { ...item, isCloned: false } };
    const wrapper = shallow(<ProjectCard {...cloneProps} />);
    wrapper.find("Button[icon='clone']").simulate("click");
    expect(cloneProps.onClone).toBeCalledTimes(1);
    expect(cloneProps.onClone).toBeCalledWith(cloneProps);
  });

  it("triggers onPull when the pull button is clicked", () => {
    const pullProps: IProps = {...props, item: { ...item, isCloned: true, isBehind: true } };
    const wrapper = shallow(<ProjectCard {...pullProps} />);
    wrapper.find("Button[icon='download']").simulate("click");
    expect(pullProps.onPull).toBeCalledTimes(1);
    expect(pullProps.onPull).toBeCalledWith(pullProps);
  });

});
