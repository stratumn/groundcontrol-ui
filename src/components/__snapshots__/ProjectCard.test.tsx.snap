// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectCard /> renders correctly 1`] = `
<Card
  className="ProjectCard"
  color="grey"
>
  <CardContent>
    <CardHeader>
      <RepositoryShortName
        repository="repository"
      />
    </CardHeader>
    <Label
      size="small"
    >
      localReferenceShort Â» remoteReferenceShort
    </Label>
    <CardDescription>
      description
    </CardDescription>
    <Divider
      horizontal={true}
    >
      <Header
        as="h6"
      >
        Latest Commits
      </Header>
    </Divider>
    <ForwardRef(Relay(CommitFeed))
      items={
        Array [
          Object {
            " $fragmentRefs": null,
            " $refType": null,
          },
          Object {
            " $fragmentRefs": null,
            " $refType": null,
          },
        ]
      }
    />
  </CardContent>
  <CardContent
    extra={true}
  >
    <div
      className="ui three buttons"
    >
      <Button
        as="button"
        color="teal"
        content="Clone"
        disabled={false}
        icon="clone"
        key="clone"
        loading={false}
        onClick={[Function]}
      />
    </div>
  </CardContent>
  <Dimmer
    active={false}
    inverted={true}
  >
    <Loader
      content="Loading project commits..."
    />
  </Dimmer>
</Card>
`;
